atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -350,6 +349,10 @@ public long setCommitIndex(final long commitIndex) {
     if (commitIndex > previousCommitIndex) {
       this.commitIndex = commitIndex;
       logWriter.commit(Math.min(commitIndex, logWriter.getLastIndex()));
+      if (raftLog.isFlushOnCommit() && isLeader()) {
+        // writer's buffer is isolated from readers
```

Comment:
I don't think this is correct - we flush on commit for the leader __because__ the leader already includes itself in the quorum, so before we can truly commit we need to persist the write.

Reply:
That's true, but the reason why we need to persist the write at all is because the buffer is not shared with the readers, right?

Reply:
commits need to be persisted - you should not lose commits otherwise your raft protocol is broken

Reply:
No, it's because the leader counts itself in the quorum as already persisted. So, let's say your quorum is 2. On a new append, the leader will write (but not flush), and so it starts its count at 1, and only needs a follower to acknowledge before achieving quorum.

So the leader is not flushing on append; what if it also didn't flush on commit? Then when the follower acknowledges the append, it would commit. Let's say this happens, and the leader now sends a new request to other nodes with the new commit index, then dies. And let's say as it dies, it hadn't flushed (since we don't know when the OS will flush automatically). Now, only one node had truly persisted the write, so we didn't actually achieve quorum.

Reply:
> commits need to be persisted - you should not lose commits otherwise your raft protocol is broken

Not quite true - the Raft paper describes that if you also wait for the async flush of the leader before counting it in the quorum, then if you still reach quorum (e.g. the 2 followers acknowledged the write before the leader flushed), then it doesn't matter, you still got quorum, so who cares if it's not flushed on the leader :man_shrugging: